### PR TITLE
Bookmarks Improvements

### PR DESCRIPTION
**Pull request by slprime**
### Save link to current recipe in bookmarks
How it work
When you add an item to the bookmarks, NEI remembers the recipe tab and recipe number. So when you select an item from bookmarks, you get into that tab and to that recipe.

Related issue
#47

How it work

Save current recipe when add item to bookmarks
This behavior can be disabled in settings
If need open item without open related recipe - use shift key and click to item
Related recipe open not only from bookmarks panel
If Recipe move, remove or add new recipes in modpack - link to the recipe work correctly or open first tab and first page

### Add Drag&Drop (sorting items) in bookmarks 
nei-bookmarks-sorting

Sorting items in bookmarks.  
For use it need hold shift

issue: https://github.com/GTNewHorizons/NotEnoughItems/issues/16

### Add possibility to extend bookmarks save/restore functions; in future supports GT++ fluids
This add possibility to extend bookmarks to works with items (save/restore & normalize)

This Request need apply after https://github.com/GTNewHorizons/NotEnoughItems/pull/173

For next step need add IStackStringifyHandler in GT++ to finish 
issue https://github.com/GTNewHorizons/NotEnoughItems/issues/18

### Add NBT support for different bookmarks
image
issue: https://github.com/GTNewHorizons/NotEnoughItems/issues/30

### CODE REVIEW

File: src/codechicken/nei/util/NBTJson.java [hunk 1]
```diff
@@ -54,9 +54,24 @@ public static JsonElement toJsonObject(NBTBase nbt) {
...
+            return new JsonPrimitive(((NBTTagInt) nbt).func_150287_d());
+        } else if (nbt instanceof NBTTagLong) {
+            // Number (long)
+            return new JsonPrimitive(Long.toString(((NBTTagLong) nbt).func_150291_c()) + 'L');
+        } else if (nbt instanceof NBTTagFloat) {
+            // Number (float)
+            return new JsonPrimitive(Float.toString(((NBTTagFloat) nbt).func_150288_h()) + 'F');
+        } else if (nbt instanceof NBTTagDouble) {
+           // Number (double)
+           return new JsonPrimitive(Double.toString(((NBTTagDouble) nbt).func_150286_g()) + 'D');
```
Comment (slprime) [inline src/codechicken/nei/util/NBTJson.java:74]: This code need for 'Add NBT support for different bookmarks' becouse `tagCompoundA.equals(tagCompoundB)` dont work correct if numbers have different type

Comment (mitchej123) [reply src/codechicken/nei/util/NBTJson.java:74]: Two things
1) Is this still backwards compatible with existing bookmarks? (Looks like it, but want to be sure)
2) Is there anything else an NBTPrimitive could be? If so add it back in at the end of the else if chain?

**Review by mitchej123 — APPROVED** [review]
Looks much better.  Holding off on merging until it's been tested on Zeta.

File: src/codechicken/nei/util/NBTJson.java (hunk 1 above)
Comment (slprime) [reply src/codechicken/nei/util/NBTJson.java:74]: 1. Yes, that is right. Exists one small problem if turn on NBT in bookmarks (#30), then maybe will not be deleted item from bookmarks if click `A` to item from another place. But from bookmarks it is deleted successful
2. Оk, I'll return it back

**Comment by bombcar** [comment]
This will fix https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/6836

**Comment by repo-alt** [comment]
Eh, will it actually be tested on Zeta if it is not merged?

**Comment by Glease** [comment]
Dream will produce a jar locally with the branch merged and ship that one

**Pull request merged**